lib/model/query/datasets.js

```diff
@@ -0,0 +1,236 @@
+
+// Returns list of Fields augmented by dataset property name OR entity creation information
+// including create, label, and id.
+const getFieldsByFormDefId = (formDefId) => ({ all }) => all(sql`
+SELECT
+  form_fields.*, ds_properties."name" as "propertyName"
+FROM
+  form_fields
+LEFT OUTER JOIN ds_property_fields ON
+  ds_property_fields."formDefId" = form_fields."formDefId"
+    AND ds_property_fields."path" = form_fields."path"
+LEFT OUTER JOIN ds_properties ON
+  ds_properties."id" = ds_property_fields."dsPropertyId"
+WHERE form_fields."formDefId" = ${formDefId}
+  AND (ds_properties."id" IS NOT NULL OR form_fields."path" LIKE '/meta/entity%')
+ORDER BY form_fields."order"
+`)
+  .then((fields) => fields.map((field) => new Form.Field(field, { propertyName: field.propertyName }))); // TODO augment this field so it also has the property name
+
+const getDatasetDiff = (projectId, formDefId) => ({ all }) => all(sql`
+  WITH form AS (
+    SELECT * FROM forms WHERE "projectId" = ${projectId} and "xmlFormId" = ${formDefId}
+  ),
+  ds AS (
+    SELECT id, name, sum(formStatus) <= 0 "isNew" FROM (
+      SELECT d.id, d.name, all_dd."formDefId", CASE WHEN all_fd."publishedAt" IS NULL THEN 0 ELSE 1 END formStatus 
+      FROM form f 
+      JOIN dataset_form_defs dd ON dd."formDefId" = f."draftDefId"  
+      JOIN datasets d ON d.id = dd."datasetId" 
+      JOIN dataset_form_defs all_dd ON all_dd."datasetId" = d.id 
+      JOIN form_defs all_fd ON all_fd.id = all_dd."formDefId"
+    ) dsInfo
+    GROUP BY id, name
+  ),
+  properties AS (
+    SELECT dp.*, dpf.*, CASE WHEN pfd."publishedAt" IS NULL THEN 0 ELSE 1 END pfStatus FROM ds_properties dp 
+    JOIN ds_property_fields dpf ON dpf."dsPropertyId" = dp.id 
+    JOIN form_defs pfd ON pfd.id = dpf."formDefId" 
+    JOIN ds ON ds.id = dp."datasetId"
+  ),
+  propertyStatus AS (
+    SELECT "datasetId", name, sum(pfStatus) <= 0 "isNew"
+    FROM properties
+    GROUP BY "datasetId", name
+  )
+  SELECT ds.name "datasetName", ds."isNew" "isDatasetNew", p.name "propertyName", p."isNew" "isPropertyNew" FROM form
+  CROSS JOIN ds
+  LEFT JOIN (
+    SELECT p.*, ps."isNew" "isNew" FROM properties p 
+    JOIN propertyStatus ps ON ps.name = p.name AND ps."datasetId" = p."datasetId" 
+  ) p ON ds.id = p."datasetId" AND form."draftDefId" = p."formDefId"
+  WHERE p.name IS NOT NULL OR ds."isNew"
+  ORDER BY p.name
+`)
+// AND ps.status = 'Draft'
+  .then(reduceBy((acc, { propertyName, isPropertyNew }) => (propertyName ? acc.concat({ name: propertyName, isNew: isPropertyNew }) : acc), [], (row) => `${row.datasetName},${row.isDatasetNew}`))
+  .then(r => Object.keys(r).map(k => ({ name: k.split(',')[0], isNew: k.split(',')[1] === 'true', properties: r[k] })));
+
+const getByProjectAndName = (projectId, name) => ({ maybeOne }) => maybeOne(sql`
+SELECT * FROM datasets WHERE "projectId" = ${projectId} AND name = ${name}`)
+  .then(map(construct(Dataset)));
```

Comment:
There is this function and there is also the following function in the code above: 

```
const getByName = (datasetName, projectId) => ({ all }) =>
  all(_getByNameSql(unjoiner(Dataset, Dataset.Property, Dataset.PropertyField).fields, datasetName, projectId))
    .then(makeHierarchy)
    .then(asArray)
    .then(nth(0));
```

They have similar names/arguments but slightly different utility.  We'll come back and give them more descriptive names later.